File: Learning/基础/二分/抽签游戏.py
def binary(x):
    l,r=0,n*n-1
    while(l<=r):
        mid=(l+r)//2
        if(b[mid]>x):r=mid-1
        elif(b[mid]<x):l=mid+1
        else:return True

def find(n,k):
    for i in range(n):
        for j in range(n):
            if(binary(k-a[i]-a[j])):return True
    return False

n,k=list(map(int,input().split()))
a=list(map(int,input().split()))
b=[]
for i in range(n):
    for j in range(n):
        b.append(a[i]+a[j])
b.sort()
if(find(n,k)):print('Yes')
else:print('No')

File: algo/Basic Algorithm/Search/Binary Search/抽签游戏.cpp
#include<bits/stdc++.h>
using namespace std;
int n,k,a[1010],b[1000010];

bool binary(int x){
	int l=1,r=n*n,mid;
	while(l<=r){
		mid=(l+r)/2;
		if(b[mid]>x) r=mid-1;
		else if(b[mid]<x) l=mid+1;
		else return true;
	}
	return false;
}

bool find(int n,int k){
	for(int i=1;i<=n;i++){
		for(int j=1;j<=n;j++){
			if(binary(k-a[i]-a[j])) return true;
		}
	}
	return false;
}

int main(){
	cin>>n>>k;
	for(int i=1;i<=n;i++){
		cin>>a[i];
	}
	for(int i=0;i<n;i++){
		for(int j=1;j<=n;j++){
			b[i*n+j]=a[i]+a[j];
		}
	}
	sort(b+1,b+n*n);
	if(find(n,k)) cout<<"Yes"<<endl;
	else cout<<"No"<<endl;
}
